cmd/crank/beta/trace/internal/resource/client.go

```diff
@@ -119,43 +121,44 @@ func (kc *Client) getResource(ctx context.Context, ref *v1.ObjectReference) *Res
 // getResourceChildrenRefs returns the references to the children for the given
 // Resource, assuming it's a Crossplane resource, XR or XRC.
 func getResourceChildrenRefs(r *Resource, getConnectionSecrets bool) []v1.ObjectReference {
-	// TODO(phisco): we could leverage annotations to infer whether the resource is an XRC, XR or MR or connection secret
 	obj := r.Unstructured
 	// collect object references for the
 	var refs []v1.ObjectReference
 
-	xrc := claim.Unstructured{Unstructured: obj}
-	if ref := xrc.GetResourceReference(); ref != nil {
-		refs = append(refs, v1.ObjectReference{
-			APIVersion: ref.APIVersion,
-			Kind:       ref.Kind,
-			Name:       ref.Name,
-			Namespace:  ref.Namespace,
-		})
-	}
-
-	xr := composite.Unstructured{Unstructured: obj}
-	refs = append(refs, xr.GetResourceReferences()...)
-
-	if !getConnectionSecrets {
-		return refs
-	}
-
-	// Handle write connection secrets for both XR and XRC
-	xrSecretRef := xr.GetWriteConnectionSecretToReference()
-	if xrSecretRef != nil {
-		ref := v1.ObjectReference{
-			APIVersion: "v1",
-			Kind:       "Secret",
-			Name:       xrSecretRef.Name,
-			Namespace:  xrSecretRef.Namespace,
+	labels := obj.GetLabels()
+	switch labels[xcrd.LabelKeyNamePrefixForComposed] {
+	case "":
+		// This is an XRC,  get the XR ref, we leave the connection secret
+		// handling to the XR
+		xrc := claim.Unstructured{Unstructured: obj}
+		if ref := xrc.GetResourceReference(); ref != nil {
+			refs = append(refs, v1.ObjectReference{
+				APIVersion: ref.APIVersion,
+				Kind:       ref.Kind,
+				Name:       ref.Name,
+				Namespace:  ref.Namespace,
+			})
+		}
+	case obj.GetName():
```

Comment:
This would hold if this is the top-level composite, but not if it is one of the nested ones. 

```yaml
kind: XNetwork
metadata:
  annotations:
    crossplane.io/composition-resource-name: compositeNetworkEKS
  generateName: platform-ref-aws-ht-ctk5x-
  labels:
    crossplane.io/claim-name: platform-ref-aws-ht
    crossplane.io/claim-namespace: default
    crossplane.io/composite: platform-ref-aws-ht-ctk5x
  name: platform-ref-aws-ht-ctk5x-6lpkx
```

Reply:
good point, rewritten as discussed, it's now working fine with nested Compositions:
```
go run github.com/crossplane/crossplane/cmd/crank beta trace nopresource.nop.example.org apiextensions-composition-functions
NAME                                                                   READY   SYNCED   STATUS
NopResource/apiextensions-composition-functions (default)              True    True     Available
└─ XNopResource/apiextensions-composition-functions-829f8              True    True     Available
   └─ XNopResource/apiextensions-composition-functions-829f8-j8wvd     True    True     Available
      └─ NopResource/apiextensions-composition-functions-829f8-mtn56   True    True
```